### PR TITLE
FF resolver support for receiving dictionary map

### PR DESCRIPTION
**Pull request by Christian862**
## Description
Please include a summary of the change, motivation and context.

### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: projects/common/src/feature/state/feature-state.resolver.ts [hunk 1]
```diff
@@ -24,3 +34,10 @@ export abstract class FeatureStateResolver {
     return FeatureState.Enabled;
   }
 }
+
+interface FeaturePageTimeRangeMap {
+  [path: string]: string;
+}
+
+export type GraphQlFeatureFlagValue = boolean | string | number | FeaturePageTimeRangeMap;
```
Comment (aaron-steinfeld) [inline projects/common/src/feature/state/feature-state.resolver.ts:42]:
```suggestion
export type FeatureFlagValue = boolean | string | number | Dictionary<string>;
```

File: projects/common/src/time/page-time-range-preference.service.ts [hunk 2]
```diff
@@ -32,10 +31,10 @@ export class PageTimeRangePreferenceService {
   public getTimeRangePreferenceForPage(rootLevelPath: string): Observable<TimeRangeResolver> {
     return combineLatest([
       this.pageTimeRangeStringDictionary$,
-      this.featureStateResolver.getFeatureState(ApplicationFeature.PageTimeRange)
+      this.featureStateResolver.getFeatureFlagValue<boolean>(ApplicationFeature.PageTimeRange)
```
Comment (aaron-steinfeld) [inline projects/common/src/time/page-time-range-preference.service.ts:34]: this should still be calling `getFeatureState`, it doesn't need the raw value

File: projects/components/src/header/application/application-header.component.ts [hunk 3]
```diff
@@ -48,8 +47,8 @@ export class ApplicationHeaderComponent {
     private readonly featureStateResolver: FeatureStateResolver
   ) {
     this.pageLevelTimeRangeDisabled$ = this.featureStateResolver
-      .getFeatureState(ApplicationFeature.PageTimeRange)
-      .pipe(map(featureState => featureState === FeatureState.Disabled));
+      .getFeatureFlagValue<boolean>(ApplicationFeature.PageTimeRange)
```
Comment (aaron-steinfeld) [inline projects/components/src/header/application/application-header.component.ts:50]: likewise here and a couple other spots in later files

File: src/app/shared/feature-resolver/feature-resolver.service.ts [hunk 4]
```diff
@@ -1,15 +1,17 @@
...
 export class FeatureResolverService extends FeatureStateResolver {
-  public getFeatureState(flag: string): Observable<FeatureState> {
-    switch (flag) {
+  public getFeatureFlagValue<T extends FeatureFlagValue = FeatureFlagValue>(feature: string): Observable<T> {
+    switch (feature) {
       case ApplicationFeature.PageTimeRange:
-        return of(FeatureState.Disabled);
+        return of(false as T);
+      case ApplicationFeature.FeatureDefaultTimeRange:
+        return of(({ 'path-1': '1h', 'path-2': '1d' } as unknown) as T); //dummy data
```
Comment (aaron-steinfeld) [inline src/app/shared/feature-resolver/feature-resolver.service.ts:12]: Should default to empty. These are really just overrides to allow config to override what's in code.

File: src/app/shared/feature-resolver/feature-resolver.service.ts [hunk 5]
```diff
@@ -1,15 +1,17 @@
...
+  public getFeatureFlagValue<T extends FeatureFlagValue = FeatureFlagValue>(feature: string): Observable<T> {
+    switch (feature) {
       case ApplicationFeature.PageTimeRange:
-        return of(FeatureState.Disabled);
+        return of(false as T);
+      case ApplicationFeature.FeatureDefaultTimeRange:
+        return of(({ 'path-1': '1h', 'path-2': '1d' } as unknown) as T); // Dummy data
       default:
-        return of(FeatureState.Enabled);
+        return EMPTY;
```
Comment (aaron-steinfeld) [inline src/app/shared/feature-resolver/feature-resolver.service.ts:14]: We should stick with previous default behavior unless a reason to switch.

File: projects/common/src/feature/state/feature-state.resolver.ts (hunk 1 above)
Comment (Christian862) [reply projects/common/src/feature/state/feature-state.resolver.ts]: fixed

File: projects/common/src/time/page-time-range-preference.service.ts (hunk 2 above)
Comment (Christian862) [reply projects/common/src/time/page-time-range-preference.service.ts]: fixed

File: projects/components/src/header/application/application-header.component.ts (hunk 3 above)
Comment (Christian862) [reply projects/components/src/header/application/application-header.component.ts]: fixed

**Review by aaron-steinfeld — APPROVED** [review]

File: src/app/shared/feature-resolver/feature-resolver.service.ts [hunk 6]
```diff
@@ -9,7 +9,7 @@ export class FeatureResolverService extends FeatureStateResolver {
       case ApplicationFeature.PageTimeRange:
         return of(false as T);
       case ApplicationFeature.FeatureDefaultTimeRange:
-        return of({} as T);
+        return of(({} as unknown) as T);
```
Comment (aaron-steinfeld) [inline src/app/shared/feature-resolver/feature-resolver.service.ts:12]: nit - don't cast to unknown. What was the issue? Either `FeatureFlagValue` is missing the type, or the fact that `{}` is empty means it's not matching the dictionary type in which case we'd rather cast it to something like `{} as Dictionary<string> & T`

Comment (Christian862) [reply src/app/shared/feature-resolver/feature-resolver.service.ts:12]: it was a tslint issue `no-object-literal-type-assertion`

Comment (Christian862) [reply src/app/shared/feature-resolver/feature-resolver.service.ts:12]: object literals can only be cast to `any` or `unknown`

Comment (aaron-steinfeld) [reply src/app/shared/feature-resolver/feature-resolver.service.ts:12]: You can either split it into declaration and usage or just ignore the lint rule. We don't want to change the casts to be less accurate to satisfy the linter, it's better to turn off the linter if we consider this an exception.

```
        const value = {};
        
        return of(value as T);
```
or
```
        // tslint:disable-next-line: no-object-literal-type-assertion
        return of({} as T);
```
would both be preferable.

**Review by aaron-steinfeld — APPROVED** [review]

**Pull request merged**